### PR TITLE
gh-116333: Relax error string text expectations in SSL-related tests

### PR DESCRIPTION
**Pull request by WillChilds-Klein**
# Description

As suggested [here][1], this change relaxes the OpenSSL error string text expectations in a number of tests. This was specifically done in support of more easily building CPython [AWS-LC][2], but because AWS-LC is a fork of [BoringSSL][3], it should increase compatibility with that library as well.

In addition to the error string relaxations, we also add some guards around the `tls-unique` channel binding being used with TLSv1.3, as that feature (described in [RFC 6929][4]) is [not defined][5] for TLSv1.3.

[1]: https://discuss.python.org/t/support-building-ssl-and-hashlib-modules-against-aws-lc/44505/4
[2]: https://github.com/aws/aws-lc
[3]: https://github.com/google/boringssl
[4]: https://datatracker.ietf.org/doc/html/rfc5929#section-3
[5]: https://datatracker.ietf.org/doc/html/rfc8446#appendix-C.5

# Call-outs

I've omitted a NEWS entry as this PR does not change anything for end users, and does not fully provide out-of-the box support for AWS-LC on its own.

# Testing
- CPython CI
- local testing with AWS-LC ([additional patch required](https://gist.github.com/WillChilds-Klein/2be18929679b0c35889e8e8bc996005f))
    + some tests still fail in local testing, but this is expected due to AWS-LC's present lack of support for [TLSv1.3 post-handshake authentication](https://www.rfc-editor.org/rfc/rfc8446.html#section-4.2.6)
- local testing with system ssl
<details>
<summary>system info</summary>
<br>
<pre>
<code>$ openssl version && uname -r && cat /etc/*release
OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
6.2.0-1014-aws
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.2 LTS"
PRETTY_NAME="Ubuntu 22.04.2 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.2 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
</code>
</pre>
</details>

* Issue: gh-116333

### CODE REVIEW

**Comment by willingc** [comment]
Since this is out of my wheelhouse, I'm adding @gpshead @encukou @alex for review as well.

A couple of general questions:
- Do we support BoringSSL currently?
- If not, does testing for "BoringSSL/AWS-LC" mean we support them?
- I'm not sure that I like the OR in the regex. Would it be better to check for the SSL version and ensure the error message matches the SSL version (OpenSSL or BoringSSL/AWS-LC)?

> local testing with AWS-LC (additional patches required, omitted here)

@WillChilds-Klein What additional patches are needed? For CPython?

**Comment by alex** [comment]
AFAIK, we do not currently have any upstream support for BoringSSL -- Google maintains that in downstream patches.

Notwithstanding that, I ping'd the BoringSSL maintainer to see if this would be helpful for them or not.

**Comment by WillChilds-Klein** [comment]
Thank you for the quick reply @willingc.

> If not, does testing for "BoringSSL/AWS-LC" mean we support them?

I don't want to speak for the CPython maintainer team, but my understanding of [the discussion thus far on Discourse](https://discuss.python.org/t/support-building-ssl-and-hashlib-modules-against-aws-lc/44505/8) is that there would be no official support for AWS-LC after this patch alone. That said, we (AWS-LC) are interested in making further changes to AWS-LC and CPython to get closer to that goal.

> I'm not sure that I like the OR in the regex. Would it be better to check for the SSL version and ensure the error message matches the SSL version (OpenSSL or BoringSSL/AWS-LC)?

This would add a few extra lines of test code per error message check, but if maintainers reach consensus that this is the way to go, I'm happy to update this PR accordingly.

Please note that we (AWS-LC) only maintain CI checks for CPython built against AWS-LC, so checking for [BoringSSL's version string](https://github.com/google/boringssl/blob/e3af7710ed006e228382c8041782cba81ff4040a/crypto/crypto.c#L209) would remain uncovered. My apologies if including `BoringSSL` in the comments confuses matters, perhaps I should remove them.

> @WillChilds-Klein What additional patches are needed? For CPython?

Yes. We (AWS-LC) maintain per-version/branch CPython patches [here](https://github.com/aws/aws-lc/tree/main/tests/ci/integration/python_patch). If this PR were to be merged in its current state, our `main` patch would look like [this](https://gist.github.com/WillChilds-Klein/2be18929679b0c35889e8e8bc996005f).

File: Lib/test/test_asyncio/test_events.py
Comment (merwok) [inline Lib/test/test_asyncio/test_events.py]: I think it would be more idiomatic to use a multi-line string here (using https://docs.python.org/3/library/re.html#re.VERBOSE)

**Comment by gpshead** [comment]
From other conversations, the BoringSSL maintainers (hi @davidben) are potentially going to make some BoringSSL changes to reduce the size of our patch set that lets CPython work with it, I'll post it as a Draft PR for reference soon (regardless of bssl changes to reduce its size having been made yet).

This _particular_ PR does things somewhat similar to what part of our BoringSSL support patch set does.  So we should be able to reconcile them all to support all of the error message forms. (Feel free to ping me if I haven't posted that by the end of the week)

File: Lib/test/test_asyncio/test_events.py
Comment (WillChilds-Klein) [reply Lib/test/test_asyncio/test_events.py]: Ah, I didn't know about the `re.X` flag. Thanks!

**Comment by merwok** [comment]
(Please avoid force pushes for Python PRs – makes it harder for reviewers to see new changes. ref: https://devguide.python.org/getting-started/pull-request-lifecycle/)

**Comment by davidben** [comment]
To be clear, as #116399 notes, that is _not_ a PR that is proposed to be merged. Our usual process on the BoringSSL side is _not_ to try to upstream patches for BoringSSL support, because we don't wish to burden upstreams. Rather, we do some combination of carrying patches locally and adding APIs for compatibility. We generally only upstream changes if we think they would also be beneficial for the project when using OpenSSL.

To that end, many of those changes are in our local copy purely because carrying them ourselves had yet to be enough of a problem (or at least we hadn't heard of it being a problem 😉) to prioritize reducing them. Every now and then I'll look around and chip away at the next one. Usually that's adding new APIs for BoringSSL, but sometimes it's fixing bugs in CPython like #95495, which was to prepare for adding `SSL_read_ex`, but in a way that doesn't break EOF, to BoringSSL.

**Comment by encukou** [comment]
@gpshead, it feels that you are blocking this PR. What are your plans with it? 

Should this PR wait for some changes in BoringSSL? Should we start merging changes common to both PRs?

Me, I'd be fine with merging the most common & trivial error message differences, as a first pass. Say:

- replacing case-and-separator mismatches like `certificate verify failed`/`CERTIFICATE_VERIFY_FAILED` with a `(?i)certificate.verify.failed` regex; 
- replacing `'PEM lib'` with `'PEM (lib|routines)'` (FWIW, Googlers, you might want to add the parens)

**Review by gpshead — APPROVED** [review]

**Comment by gpshead** [comment]
BTW, I like your use of multi-line regular expressions in this PR.  that's cleaner than our minimize-the-diff-size long strings in our existing BoringSSL support changes.

**Pull request merged**
